core/src/connection.rs

```diff
@@ -231,68 +233,122 @@ where
+    /// complete, `poll()` returns `Ok(None)`.
+    pub fn start_close(&mut self) {
+        if self.state == ConnectionState::Open {
+            self.state = ConnectionState::CloseHandler;
+        }
     }
 
-    /// Begins an orderly shutdown of the connection, returning a
-    /// `Future` that resolves when connection shutdown is complete.
-    pub fn close(self) -> Close<TMuxer> {
-        self.muxing.close().0
+    /// Whether the connection is open, i.e. neither closing nor already closed.
+    pub fn is_open(&self) -> bool {
+        self.state == ConnectionState::Open
     }
 
     /// Polls the connection for events produced by the associated handler
     /// as a result of I/O activity on the substream multiplexer.
+    ///
+    /// > **Note**: A return value of `Ok(None)` signals successful
+    /// > connection shutdown, whereas an `Err` signals termination
+    /// > of the connection due to an error. In either case, the
+    /// > connection must be dropped; any further method calls
+    /// > result in unspecified behaviour.
     pub fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>)
-        -> Poll<Result<Event<THandler::OutEvent>, ConnectionError<THandler::Error>>>
+        -> Poll<Result<Option<Event<THandler::OutEvent>>, ConnectionError<THandler::Error>>>
     {
         loop {
+            if let ConnectionState::Closed = self.state { // (1)
+                return Poll::Ready(Ok(None))
+            }
+
+            if let ConnectionState::CloseMuxer = self.state { // (2)
+                match futures::ready!(self.muxing.poll_close(cx)) {
+                    Ok(()) => {
+                        self.state = ConnectionState::Closed;
+                        return Poll::Ready(Ok(None))
+                    }
+                    Err(e) => {
+                        return Poll::Ready(Err(ConnectionError::IO(e)))
+                    }
+                }
+            }
+
+            // At this point the connection is either open or in the process
+            // of a graceful shutdown by the connection handler.
             let mut io_pending = false;
 
             // Perform I/O on the connection through the muxer, informing the handler
-            // of new substreams.
+            // of new substreams or other muxer events.
             match self.muxing.poll(cx) {
                 Poll::Pending => io_pending = true,
                 Poll::Ready(Ok(SubstreamEvent::InboundSubstream { substream })) => {
-                    self.handler.inject_substream(substream, SubstreamEndpoint::Listener)
+                    // Drop new inbound substreams when closing. This is analogous
+                    // to rejecting new connections.
+                    if self.state == ConnectionState::Open {
+                        self.handler.inject_substream(substream, SubstreamEndpoint::Listener)
+                    } else {
+                        log::trace!("Inbound substream dropped. Connection is closing.")
```

Comment:
Yes, the current idea is to treat new inbound substreams analogously (as much as possible anyway) to new connections during shutdown - that is, to refuse them at the earliest possible moment. Any failures relating to attempts at creating new substreams or connections towards the peer that is shutting down should ideally result in retries with a different peer, much like without the graceful shutdown.

Reply:
I am not sure if the shutdown logic as currently implemented can really alleviate error rates because any error in one substream will often lead to a connection close, causing failures in other substreams. For example, the request-response handler will immediately close the connection when any inbound or outbound upgrade fails other than via timeout or protocol mismatch. This means that if the remote closes the connection and my next outbound request over this connection fails, all my ongoing requests to this connection will fail too.

Reply:
That is true, unfortunately. It depends very much on how substream write errors are handled, in the specific case of `libp2p-request-response` by `RequestResponseCodec::write_request`, which is currently completely up to the user. As you say, most other protocols probably consider write errors on a substream as fatal as well. The only somewhat clean solution that comes to my mind would seem to lead back to the ability for the substream multiplexers to have a configurable inbound substream limit similar to that of the connection limit for the `Pool` which can be changed during shutdown (i.e. set to 0), i.e. a limit that does not affect existing substreams but can be used to exert back-pressure in a way such that the remote knows it reached the current substream limit on that connection, but the connection itself is not broken.